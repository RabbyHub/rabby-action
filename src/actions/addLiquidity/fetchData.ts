import PQueue from 'p-queue';
import {
  AddLiquidityRequireData,
  FetchActionRequiredData,
  ParsedActionData,
} from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import BigNumber from 'bignumber.js';

export const fetchDataAddLiquidity: FetchActionRequiredData<{
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: any;
  token1: any;
}> = async (options, likeAction) => {
  const { actionData, apiProvider, sender, walletProvider, chainId } = options;

  const action =
    likeAction || (<ParsedActionData<'transaction'>>actionData).addLiquidity;

  if (!action || options.type === 'text') {
    return {};
  }

  const { receiver, exchange_rate, token0, token1 } = action;

  const fair_exchange_rate = new BigNumber(token0.price).div(token1.price);

  const diff = +new BigNumber(exchange_rate)
    .minus(fair_exchange_rate)
    .abs()
    .div(fair_exchange_rate)
    .times(100)
    .toFixed(4);

  const receiverInWallet = await walletProvider.hasAddress(receiver);

  const id =
    options.type === 'transaction'
      ? options.tx.to
      : options.actionData.contractId;
  if (!id) {
    return {};
  }

  const result: AddLiquidityRequireData = {
    id,
    protocol: null,
    bornAt: 0,
    rank: null,
    sender,
    receiverInWallet,
    hasInteraction: false,
    receiver,
    poolRate: exchange_rate,
    token0,
    token1,
    marketRate: fair_exchange_rate.toFixed(),
    diff,
  };
  const queue = new PQueue();

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(id, chainId);
    if (!contractInfo) {
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.bornAt = contractInfo.create_at;

      result.protocol = contractInfo.protocol;
    }
  });

  queue.add(async () => {
    const hasInteraction = await apiProvider.hasInteraction(
      sender,
      chainId,
      id
    );
    result.hasInteraction = hasInteraction.has_interaction;
  });

  await waitQueueFinished(queue);

  return result;
};
