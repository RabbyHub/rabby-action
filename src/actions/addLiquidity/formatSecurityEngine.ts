import {
  AddLiquidityRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineAddLiquidity: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.addLiquidity || !chainId) {
      return {} as any;
    }

    const data = requireData as AddLiquidityRequireData;

    return {
      addLiquidity: {
        receiverMatch:
          data.receiver?.toLowerCase() === data.sender.toLowerCase(),
        diff: data.diff,
      },
    };
  };
