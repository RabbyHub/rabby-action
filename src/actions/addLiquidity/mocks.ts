export const preExecData = {
  abi: {
    func: 'mint',
    params: [
      [
        '0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22',
        '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913',
        10,
        -4970,
        -4960,
        2955759909,
        885199302,
        2940981110,
        880773306,
        '0xE2866F2291B5c5bafD2858B9247f93745108A523',
        1754366820,
        0,
      ],
    ],
  },
  abi_str:
    '{\n    "func": "mint",\n    "params": [\n        [\n            "0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22",\n            "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",\n            10,\n            -4970,\n            -4960,\n            2955759909,\n            885199302,\n            2940981110,\n            880773306,\n            "0xE2866F2291B5c5bafD2858B9247f93745108A523",\n            1754366820,\n            0\n        ]\n    ]\n}',
  balance_change: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
    send_token_list: [],
    receive_token_list: [],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: 0,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 29320,
    gas_limit: 0,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'base',
    chain: 'base',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 3629.55,
    price_24h_change: 0.018385589869839923,
    credit_score: 46503412.60907772,
    is_verified: true,
    is_scam: false,
    is_suspicious: null,
    is_core: true,
    is_wallet: true,
    time_at: null,
    low_credit_score: false,
    amount: 1e-5,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
  },
  type_call: {
    action: 'mint',
    contract: '0x827922686190790b37229fd06084350e74485b72',
    contract_protocol_name: 'Aerodrome V3',
    contract_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/base_aerodrome/f02d753bc321dc8ba480f0424a686482.png',
  },
  trace_id: '663ae6c6c0ac4b1aa5651fadba316de2',
} as any;

export const parseTxData = {
  action: { type: 'cancel_tx' },
  contract_call: null,
  log_id: 441860918,
} as any;

export const txData = {
  chainId: 42161,
  from: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  to: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  value: '0x0',
  data: '0x',
  gas: '0x0',
  maxFeePerGas: '0xb71b00',
  maxPriorityFeePerGas: '0xb71b00',
  nonce: '0x4',
};
