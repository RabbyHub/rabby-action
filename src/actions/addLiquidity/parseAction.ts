import { ParseAction } from '../../types';
import { TokenItem } from '@rabby-wallet/rabby-api/dist/types';

interface AddLiquidityActionData {
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: TokenItem;
  token1: TokenItem;
}

export const parseActionAddLiquidity: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'add_liquidity') {
    return {};
  }

  const {
    receiver,
    exchange_rate,
    min_exchange_rate,
    max_exchange_rate,
    token0,
    token1,
  } = data.data as unknown as AddLiquidityActionData;

  return {
    addLiquidity: {
      receiver,
      exchange_rate,
      min_exchange_rate,
      max_exchange_rate,
      token0,
      token1,
    },
  };
};
