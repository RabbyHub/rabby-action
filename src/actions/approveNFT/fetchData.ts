import PQueue from 'p-queue';
import { ApproveNFTRequireData, FetchActionRequiredData } from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import { fetchHasInteraction } from '../../fetches/fetchHasInteraction';

export const fetchDataApproveNFT: FetchActionRequiredData<{
  spender: string;
}> = async (options, likeAction) => {
  const queue = new PQueue();
  const { sender, apiProvider, chainId, actionData, extraActionDataState } =
    options;
  const action = likeAction || actionData.approveNFT;

  if (!action) {
    return {};
  }

  const result: ApproveNFTRequireData = {
    isEOA: false,
    contract: null,
    riskExposure: 0,
    rank: null,
    bornAt: 0,
    protocol: null,
    isDanger: false,
    tokenBalance: '0',
    hasInteraction: null,
    extraState: {
      hasInteraction: () => undefined,
    },
  };

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(
      action.spender,
      chainId
    );
    if (!contractInfo) {
      result.isEOA = true;
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.riskExposure = contractInfo.top_nft_spend_usd_value;
      result.bornAt = contractInfo.create_at;
      result.isDanger =
        contractInfo.is_danger.auto || contractInfo.is_danger.edit;
      result.protocol = contractInfo.protocol;
    }

    if (result.isEOA) {
      queue.add(async () => {
        const { desc } = await apiProvider.addrDesc(action.spender);
        result.bornAt = desc.born_at;
      });
    }
  });

  fetchHasInteraction({
    apiProvider,
    sender,
    chainId,
    spender: action.spender,
    extraActionDataState,
    queue,
    result,
  });

  await waitQueueFinished(queue);
  return result;
};
