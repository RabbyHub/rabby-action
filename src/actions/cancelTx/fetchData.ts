import { FetchActionRequiredData } from '../../types';

export const fetchDataCancelTx: FetchActionRequiredData = async (options) => {
  const { walletProvider, actionData, chainId, sender } = options;

  if (!actionData.cancelTx) {
    return {};
  }

  const chain = walletProvider.findChain({
    serverId: chainId,
  });
  if (chain && actionData.cancelTx.nonce !== undefined) {
    const pendingTxs = await walletProvider.getPendingTxsByNonce(
      sender,
      chain.id,
      Number(actionData.cancelTx.nonce)
    );
    return {
      pendingTxs,
    };
  } else {
    return {
      pendingTxs: [],
    };
  }
};
