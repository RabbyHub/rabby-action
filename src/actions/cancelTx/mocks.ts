export const preExecData = {
  abi: null,
  abi_str: null,
  balance_change: {
    success: true,
    error: null,
    send_token_list: [],
    receive_token_list: [],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: 0,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 21000,
    gas_limit: 21000,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'eth',
    chain: 'eth',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 2683.8,
    price_24h_change: 0.022289342479632213,
    credit_score: 100000000.0,
    is_verified: true,
    is_scam: false,
    is_suspicious: false,
    is_core: true,
    is_wallet: true,
    time_at: 1483200000.0,
    amount: 0.00336741278341931,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: true,
    error: null,
  },
  type_cancel_tx: {},
  trace_id: '60184ce4e78b444e9d55e8201ac883e2',
} as any;

export const parseTxData = {
  action: { type: 'cancel_tx' },
  contract_call: null,
  log_id: 85276307,
} as any;

export const txData = {
  chainId: 1,
  from: '0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85',
  to: '0x5853eD4f26A3fceA565b3FBC698bb19cdF6DEB85',
  value: '0x0',
  data: '0x',
  gas: '0x5208',
  maxFeePerGas: '0x4190ab00',
  maxPriorityFeePerGas: '0x4190ab00',
  nonce: '0x1691',
};
