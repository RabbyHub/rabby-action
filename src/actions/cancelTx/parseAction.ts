import { ParseAction } from '../../types';

export const parseActionCancelTx: ParseAction<'transaction'> = (options) => {
  const { data } = options;

  if (data?.type !== 'cancel_tx') {
    return {};
  }

  return {
    cancelTx: {
      nonce: options.tx.nonce,
    },
  };
};
