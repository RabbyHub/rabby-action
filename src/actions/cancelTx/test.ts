import {
  apiProvider,
  ETH_CHAIN_ID,
  formatProvider,
  SENDER,
  walletProvider,
} from '../../../__mocks__';
import { ParsedTransactionActionData } from '../../types';
import { parseTxData, preExecData, txData } from './mocks';
import { parseActionCancelTx } from './parseAction';
import { fetchDataCancelTx } from './fetchData';
import { formatSecurityEngineCancelTx } from './formatSecurityEngine';
import {
  parseAction,
  fetchActionRequiredData,
  formatSecurityEngineContext,
} from '../..';

/**
 * SENDER send 0 ETH to SENDER
 */
test.each([
  [parseActionCancelTx, fetchDataCancelTx, formatSecurityEngineCancelTx],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])('CancelTx', async (_parseAction, _fetchData, _format) => {
  const actionData = _parseAction({
    type: 'transaction',
    data: parseTxData['action'],
    balanceChange: preExecData.balance_change,
    sender: txData.from,
    preExecVersion: preExecData.pre_exec_version,
    gasUsed: preExecData.gas.gas_used,
    tx: txData,
  }) as ParsedTransactionActionData;
  expect(actionData).toMatchSnapshot('parseActionCancelTx');

  const requireData = await _fetchData({
    type: 'transaction',
    actionData,
    contractCall: parseTxData.contract_call,
    chainId: ETH_CHAIN_ID,
    sender: SENDER,
    walletProvider,
    tx: txData,
    apiProvider,
  });
  expect(requireData).toMatchSnapshot('fetchDataCancelTx');

  const ctx = await _format({
    type: 'transaction',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineCancelTx');
});
