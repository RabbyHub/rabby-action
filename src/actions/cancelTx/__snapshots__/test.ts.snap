// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CancelTx: fetchDataCancelTx 1`] = `
{
  "pendingTxs": [],
}
`;

exports[`CancelTx: fetchDataCancelTx 2`] = `
{
  "pendingTxs": [],
}
`;

exports[`CancelTx: formatSecurityEngineCancelTx 1`] = `{}`;

exports[`CancelTx: formatSecurityEngineCancelTx 2`] = `{}`;

exports[`CancelTx: parseActionCancelTx 1`] = `
{
  "cancelTx": {
    "nonce": "0x1691",
  },
}
`;

exports[`CancelTx: parseActionCancelTx 2`] = `
{
  "cancelTx": {
    "nonce": "0x1691",
  },
}
`;
